cmd/refactor.go

```diff
@@ -0,0 +1,177 @@
+
+Example:
+--------
+
+"policy.rego" contains the below policy:
+ _ _ _ _ _ _ _ _ _ _ _ _ _
+| package lib.foo         |
+|                         |
+| default allow = false   |
+| _ _ _ _ _ _ _ _ _ _ _ _ |     
+	
+	$ opa refactor move -p data.lib.foo:data.baz.bar policy.rego
+
+The 'move' command outputs the below policy to stdout with the package name rewritten as per the mapping:
+
+ _ _ _ _ _ _ _ _ _ _ _ _ _
+| package baz.bar         |
+|                         |
+| default allow = false   |
+| _ _ _ _ _ _ _ _ _ _ _ _ | 
+`,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return validateMoveArgs(args)
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := doMove(moveCommandParams, args, os.Stdout); err != nil {
+				fmt.Fprintln(os.Stderr, "error:", err)
+				os.Exit(1)
+			}
+		},
+	}
+
+	moveCommand.Flags().VarP(&moveCommandParams.mapping, "path", "p", "set the mapping that defines how references should be rewritten (ie. <from>:<to>). This flag can be repeated.")
+	moveCommand.Flags().BoolVarP(&moveCommandParams.overwrite, "write", "w", false, "overwrite the original source file")
+	addIgnoreFlag(moveCommand.Flags(), &moveCommandParams.ignore)
+	refactorCommand.AddCommand(moveCommand)
+	RootCommand.AddCommand(refactorCommand)
+}
+
+func doMove(params moveCommandParams, args []string, out io.Writer) error {
+	if len(params.mapping.v) == 0 {
+		return errors.New("specify at least one mapping of the form <from>:<to>")
+	}
+
+	srcDstMap, err := parseSrcDstMap(params.mapping.v)
+	if err != nil {
+		return err
+	}
+
+	modules := map[string]*ast.Module{}
+
+	f := loaderFilter{
+		Ignore: params.ignore,
+	}
+
+	result, err := loader.NewFileLoader().Filtered(args, f.Apply)
+	if err != nil {
+		return err
+	}
+
+	for _, m := range result.Modules {
+		modules[m.Name] = m.Parsed
+	}
+
+	movedModules, err := refactor.New().Move(refactor.MoveQuery{
+		Modules:       modules,
+		SrcDstMapping: srcDstMap,
+	})
+	if err != nil {
+		return err
+	}
+
+	for filename, mod := range movedModules.Result {
+		filename, err = fileurl.Clean(filename)
+		if err != nil {
+			return err
+		}
+
+		formatted, err := format.Source(filename, []byte(mod.String()))
+		if err != nil {
+			return newError("failed to parse Rego source file: %v", err)
+		}
+
+		if params.overwrite {
+			info, err := os.Stat(filename)
+			if err != nil {
+				return err
+			}
+
+			outfile, err := os.OpenFile(filename, os.O_WRONLY|os.O_TRUNC, info.Mode())
+			if err != nil {
+				return newError("failed to open file for writing: %v", err)
+			}
+			defer outfile.Close()
+			out = outfile
+		}
+
+		_, err = out.Write(formatted)
+		if err != nil {
+			return newError("failed writing formatted contents: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func parseSrcDstMap(data []string) (map[string]string, error) {
+	result := map[string]string{}
+
+	for _, d := range data {
+		parts := strings.Split(d, ":")
```

Comment:
I haven't fully checked but does this mean we cannot move stuff like `package t["foo:bar"].baz`?

I suppose one way out, if we're eager to support this, would be to either introduce the possibility to escape chars, or use some other input format for the mapping, like JSON of some sort. (A third option could be positional arguments, treating the from-to pairs as two subsequent strings, and we'd piggy-back on the bash escaping rules to provide the option for spaces, colons, etc..)

Reply:
```rego
package t["foo:bar"].baz
q {
    data.foo["some:thing"].baz
}
```

If we did `opa refactor move -p data.foo:data.fox -p data.t:data.acme.hidden policy.rego`

We'd get 
```
package acme.hidden["foo:bar"].baz
q {
	data.fox["some:thing"].baz
}
```

And for a policy like

```rego
package t["foobar"].baz
q {
    data.foo["some:thing"].baz
}
```
If we did `opa refactor move -p data.foo:data.fox -p data.t.foobar:data.acme.hidden.baxbux policy.rego`

We’d get

```rego
package acme.hidden.baxbux.baz
q {
	data.fox["some:thing"].baz
}
```

Reply:
Thanks for the examples. Could we move from `data.foo["some:thing"].bar` to `data.fox.whatever.baz`?

Reply:
Not currently. I'll create an issue to track this.